File: ruby/lesson 7/is_valid_bst.rb
# https://leetcode.com/problems/validate-binary-search-tree/

def is_valid_bst(root)
    is_valid(root)
end

def is_valid(node, min = nil, max = nil)
    return true if node.nil?
    return false if !min.nil? && node.val <= min || !max.nil? && node.val >= max
    return is_valid(node.left, min, node.val) && is_valid(node.right, node.val, max)
end

File: python/lesson 7/is_valid_bst.py
# https://leetcode.com/problems/validate-binary-search-tree/

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.isValid(root)
        
    def isValid(self, node: Optional[TreeNode], min: Optional[int]=None, max: Optional[int]=None) -> bool:
        if node == None: return True
        if min != None and node.val <= min or max != None and node.val >= max: return False
        return self.isValid(node.left, min, node.val) and self.isValid(node.right, node.val, max)